### PR TITLE
Eng 13411

### PR DESCRIPTION
**Pull request by manjujames**
- New attribute 'internal' on <ssl> to indicate that inter-node communication must use SSL as well
- Changed our SSL classes to work with PicoNetwork as well and use them from PicoNetwork depending on the configuration.

### CODE REVIEW

File: src/frontend/org/voltcore/utils/ssl/MessagingChannel.java [hunk 1]
```diff
@@ -32,6 +33,7 @@
  * if SSL is enabled.
  */
 public class MessagingChannel {
+    private static final VoltLogger hostLog = new VoltLogger("HOST");
```
Comment (zeemanhe) [inline src/frontend/org/voltcore/utils/ssl/MessagingChannel.java]: **hostLog** is unused.

File: src/frontend/org/voltcore/messaging/HostMessenger.java [hunk 2]
```diff
@@ -913,7 +937,8 @@ private Integer selectNewHostId(String address) throws Exception {
     private void writeRequestJoinResponse(
             int hostId,
             JoinAcceptor.PleaDecision decision,
-            SocketChannel socket) throws Exception {
+            SocketChannel socket,
+            MessagingChannel messagingChannel) throws Exception {
```
[the remark this answers is not in the document]
Comment (manjujames) [reply src/frontend/org/voltcore/messaging/HostMessenger.java]: No, messaging channel does not have a reference to the socket.

File: src/frontend/org/voltcore/messaging/SocketJoiner.java [hunk 3]
```diff
@@ -119,6 +127,8 @@ public void notifyOfJoin(
          */
         public void requestJoin(
                 SocketChannel socket,
```
[the remark this answers is not in the document]
Comment (manjujames) [reply src/frontend/org/voltcore/messaging/SocketJoiner.java]: No, it doesn't have a reference to the socket.

File: src/frontend/org/voltcore/messaging/SocketJoiner.java [hunk 4]
```diff
@@ -405,40 +420,75 @@ private void doBind() throws Exception {
...
+            throw new IOException("Rejected accepting new internal connection, SSL handshake failed: " + e.getMessage(), e);
         }
-        messageBytes.flip();
-
-        JSONObject jsObj = new JSONObject(new String(messageBytes.array(), StandardCharsets.UTF_8));
-        return jsObj;
+        if (!handshakeStatus) {
+            try {
+                sc.close();
+            } catch (IOException e) {
```
[the remark this answers is not in the document]
Comment (manjujames) [reply src/frontend/org/voltcore/messaging/SocketJoiner.java]: Yes, I agree, that makes more sense.

File: src/frontend/org/voltcore/network/TLSEncryptionAdapter.java [hunk 5]
```diff
@@ -153,8 +164,98 @@ void waitForPendingEncrypts() throws IOException {
...
+                continue;
+            }
+
+            final int partialSize = m_partialSize;
+            if (partialSize > 0) {
+                assert frame.chunks == partialSize + 1
+                        : "partial frame buildup has wrong number of preceding pieces";
+
+                //TODO: Review - I don't think this synchronized(m_partialMessages) is required.
+                // See comment above.
```
[the remark this answers is not in the document]
Comment (manjujames) [reply src/frontend/org/voltcore/network/TLSEncryptionAdapter.java]: Just to be safe, I will take out the synchronized after 8.0 is out.

File: src/frontend/org/voltdb/compiler/DeploymentFileSchema.xsd [hunk 6]
```diff
@@ -332,6 +332,7 @@
     <xs:attribute name="enabled" type="xs:boolean" default="false"/>
     <xs:attribute name="external" type="xs:boolean" default="false"/>
     <xs:attribute name="dr" type="xs:boolean" default="false"/>
+    <xs:attribute name="internal" type="xs:boolean" default="false"/>
```
[the remark this answers is not in the document]
Comment (manjujames) [reply src/frontend/org/voltdb/compiler/DeploymentFileSchema.xsd]: I don't think we give any errors. This configuration existed for a while now. We initialize the SSL context only once, which means that deployment changes to SSL will not take effect even if we gave no error. I will open a bug about this.

**Pull request merged**